sway/tree/layout.c

```diff
@@ -166,9 +188,186 @@ void container_move_to(struct sway_container *container,
+				if (target->children->items[dest_i] == container) {
+					return;
+				}
+				break;
+			}
+			struct sway_output *next_output = next->data;
+			current = target = next_output->swayc;
+			wlr_log(L_DEBUG, "Selected next output (%s)", current->name);
+			break;
+		} else if (current->type == C_WORKSPACE) {
+			if (is_perpendicular(current->layout, dir)) {
+				wlr_log(L_DEBUG, "Moving perpendicular to workspace");
+				break;
+			} else {
+				// Just skip to the output
+				current = target = current->parent;
+			}
+		} else if (!is_perpendicular(current->parent->layout, dir) &&
+				offs_i >= 0 && offs_i < current->parent->children->length) {
+			target = current->parent->children->items[offs_i];
+			wlr_log(L_DEBUG, "Selected sibling %p %s '%s'", target,
+					container_type_to_str(target->type), target->name);
+			break;
+		} else {
+			current = target = current->parent;
+			wlr_log(L_DEBUG, "Selected parent %p %s '%s'", current,
+					container_type_to_str(current->type), current->name);
+		}
+	}
+
+	wlr_log(L_DEBUG, "target: %p %s '%s'", target,
+			container_type_to_str(target->type), target->name);
+
+	struct sway_container *parent = container->parent;
+	switch (target->type) {
+	case C_VIEW:
+		if (target->parent == container->parent) {
+			wlr_log(L_DEBUG, "Swapping siblings");
+			target->parent->children->items[i + offs] = container;
+			target->parent->children->items[i] = target;
+			arrange_windows(target->parent, -1, -1);
+		} else {
+			wlr_log(L_DEBUG, "Reparenting container");
+			container_add_sibling(target, container);
+			container->width = container->height = 0;
+			arrange_windows(target->parent, -1, -1);
+			arrange_windows(parent, -1, -1);
+		}
+		break;
+	case C_OUTPUT: {
+		target = seat_get_focus_inactive(
+				config->handler_context.seat, target);
+		if (target->type != C_WORKSPACE) {
+			target = container_parent(target, C_WORKSPACE);
+		}
+		container_move_to(container, target);
+		break;
+	}
+	case C_WORKSPACE: {
+		wlr_log(L_DEBUG, "Moving to workspace and reparenting others");
+		parent = container_split(target, target->layout);
+		container_remove_child(container);
+		for (int i = 0; i < target->children->length; ++i) {
+			struct sway_container *child = target->children->items[i];
+			container_move_to(parent, child);
+		}
+		int dest_i = offs > 0 ? target->children->length : 0;
+		wlr_log(L_DEBUG, "%d %d", offs, dest_i);
+		container_insert_child(target, container, dest_i);
+		container_set_layout(target,
+				dir == MOVE_LEFT || dir == MOVE_RIGHT ?  L_HORIZ : L_VERT);
+		arrange_windows(target, -1, -1);
+		break;
+	}
+	case C_CONTAINER:
+		if (target->children->length == 0) {
+			container_move_to(container, target);
+			arrange_windows(target->parent, -1, -1);
+			arrange_windows(parent, -1, -1);
+		} else if (!is_perpendicular(target->layout, dir)) {
+			// Insert in nearest side in direction of movement
+			int dest_i = offs < 0 ? target->children->length - 1 : 0;
```

Comment:
I think this should be `target->children->length` instead of `target->children->length - 1`, just like at line 334.

Reply:
ack, this was causing a bug